### PR TITLE
Add testing

### LINKED ISSUE #1: Bug report: round_sig(x, sig) not rounding correctly for some (even possibly) numbers ending with 5
The function round_sig(x, sig) is not rounding some numbers correctly when they have 5 as the last digit. This is due to the rounding half to even principle that the sigfig.round() function uses. The round_sig() function calls a wrapper function in problem_bank_helpers module called roundp() which in turn calls the sigfig.round() function which is the probable cause of this bug.

To reproduce: 

```
pbh.round_sig(25,1)
pbh.round_sig(45,1)
pbh.round_sig(85,1)
pbh.round_sig(4.675, 3)
```

Generated Output: 
20
40
80
4.67

Desired Output:
30
50
90
4.68

Tests that fail due to this bug:
```
test_round_sig_numbers_ending_with_n() # for n=5
test_round_sig_edge_cases()
```

### LINKED ISSUE #2: Bug report: sigifgs(x) does not return intended value for eFormatted strings
The function sigfigs(x) accepts only strings as a parameter and takes into account strings formatted as 1e10 and then returns the number of sigfigs in the number. But it does not produce the intended output for the strings formatted as 1e10 etc. 

To reproduce:

```
pbh.sigfigs("1e10")
```

Generated output:
0

Expected output:
1

Tests that fail due to this bug: 
```
test_sigfigs_eFormat()
```

Note: The function works when we pass in 1.e10 instead of 1e10.

### LINKED ISSUE #3: Bug report: sigfigs(x) does not return a number more than 9
The function sigfigs(x) does not return a number more than 9 even when the significant digits in the number passed in is more than 9. The maximum it returns is 9.

To reproduce:
```
pbh.sigfigs("1.234567891") 
pbh.sigfigs("3.141592653")
pbh.sigfigs("-3.1415926535")
pbh.sigfigs("123456789101112")
```

Generated output:
9
9
9
9

Expected output:
10
10
11
15

Tests that fail due to this bug:
```
test_sigfigs_between0and1()
test_sigfigs_negative()
test_sigfigs_upto12dp()
```

### LINKED ISSUE #10: problem with pbh.roundp(): kernel takes a long time to execute
This code takes a while to run and execute, seems like this is a problem with the code, but it's not obvious why.

```
import problem_bank_helpers as pbh
pbh.roundp(math.pi * (d * 10 ** (-6) / 2) ** 2, sigfigs=10)
```

### LINKED ISSUE #11: roundp breaking with large numbers
roundp breaks the scripts that parses the code when the arguments are very large (greater than 10^23)

[https://github.com/open-resources/instructor_physics_bank/pull/1178]

It can be seen in this branch how the script never exits, throwing no errors just does not finish executing.

### LINKED ISSUE #13: Add tests for problem_bank_helpers
Here is the lists of tests that we need to add:

- check rounding of numbers correctly
- check handling of trailing and leading 0s in numbers
- check that the csv files are loaded in correctly
- ... more to come soon

### PR DESCRIPTION
**Pull request by amurph24**
This branch adds testing for the rounding and sigfig functions in (problem_bank_helpers.py)[src/problem_bank_helpers/problem_bank_helpers.py] in the rounding tests.
These rounding tests check the outputs for common and edge cases, and check the input and output parameter type of the functions.
It would be beneficial to have another check on the `num_as_str` test, to ensure the test values are correct

At the moment the other functionality tests check that:
- each csv file can be loaded in using `pandas.read_csv`, contain readable data and have only 1 column (can/should be removed or changed). There is also a check that this test fails if it opens an empty csv file.
- the dataframe produced by each csv file does not contain any null or empty values (this also might need to be changed)
I am unsure of the exact use for these files, other than randomizing variables, so I have kept the tests vague for the moment.

The tests can be run from the root dir with:
```
pytest tests/
```
and perform as seen below at this point in time (18/06/2023):
<details>
      <summary>test short summary</summary>
      ```
      
      =========================== short test summary info ===========================
      FAILED tests/test_problem_bank_helpers_other_functionality.py::test_load_csv[data/metals.csv]
      FAILED tests/test_problem_bank_helpers_rounding.py::test_sigfigs[between0and1]
      FAILED tests/test_problem_bank_helpers_rounding.py::test_sigfigs[eFormat] - A...
      FAILED tests/test_problem_bank_helpers_rounding.py::test_num_as_str[1.555-2-1.56]
      FAILED tests/test_problem_bank_helpers_rounding.py::test_num_as_str[-1.5-2--2.00]
      FAILED tests/test_problem_bank_helpers_rounding.py::test_num_as_str[1234567890.1234567-8-1234567890.12345679]
      FAILED tests/test_problem_bank_helpers_rounding.py::test_num_as_str[987.654321--2-1000.00]
      FAILED tests/test_problem_bank_helpers_rounding.py::test_roundp - AssertionEr...
      FAILED tests/test_problem_bank_helpers_rounding.py::test_roundstr - Assertion...
      ======================== 9 failed, 42 passed in 5.13s =========================
      
      ```
</details>

### CODE REVIEW

File: tests/test_problem_bank_helpers_rounding.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    assert pbh.roundp('99.8765') == '100'
+
+    """Test rounding a float with format='std'"""
+    assert pbh.roundp(0.123, format='std') == 0.12
+
+    """Test rounding a float with format='English'"""
+    assert pbh.roundp(12345.6789, format='English') == 12300.0
+
+    """Test rounding a float with format='sci'"""
+    assert pbh.roundp(0.000123, format='sci') == '1.20e-04'
```
Comment (Bluesy1) [inline tests/test_problem_bank_helpers_rounding.py]:
```suggestion
def test_roundp_with_overriden_sigfig_settings():
    """Test rounding a float with specified sigfigs"""
    assert pbh.roundp(123.456, sigfigs=2) == 120.0

def test_roundp_with_overriden_decimal_settings():
    """Test rounding an integer with specified decimals"""
    assert pbh.roundp(987, decimals=2) == 987.0

def test_roundp_with_default_settings():
    """Test rounding a string with default sigfigs"""
    assert pbh.roundp('99.8765') == '100'

def test_roundp_with_format_std():
    """Test rounding a float with format='std'"""
    assert pbh.roundp(0.123, format='std') == 0.12

def test_roundp_with_format_english():
    """Test rounding a float with format='English'"""
    assert pbh.roundp(12345.6789, format='English') == 12300.0

def test_roundp_format_sci():
    """Test rounding a float with format='sci'"""
    assert pbh.roundp(0.000123, format='sci') == '1.20e-04'
```
Suggestion: Separate these into separate tests, so if the function ever changes or regresses from expected state, we can see multiple failures in a single test run.

File: tests/test_problem_bank_helpers_rounding.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    assert pbh.round_str(12345.6789, format='English') == 12300.0
+
+    """Test rounding a float with specified format='sci'"""
+    assert pbh.round_str(0.000123, format='sci') == '1.20e-04'
+
+    """Test rounding a float with specified sigfigs and format='std'"""
+    assert pbh.round_str(0.123, sigfigs=3, format='std') == 0.120
+
+    """Test rounding an integer with specified decimals and format='sci'"""
+    assert pbh.round_str(987, decimals=2, format='sci') == '9.87e+02'
```
Comment (Bluesy1) [inline tests/test_problem_bank_helpers_rounding.py]: Same thing as with `test_roundp` - consider splitting it up into multiple tests

File: tests/test_problem_bank_helpers_rounding.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    return inputs
+
+# Test sigfigs function
+@pytest.mark.parametrize(
+    "test_group",
+    group_names
+)
+def test_sigfigs(test_group, variables):
+    for i in range (0, len(variables[test_group])):
+        test_input = str(variables[test_group][i])
```
Comment (Bluesy1) [inline tests/test_problem_bank_helpers_rounding.py]:
```suggestion
    for i, test_input in enumerate(variables[test_group]):
        test_input = str(test_input)
```
Suggestion: use `enumerate` over excessive indexing since we want both the item index and item

File: tests/test_problem_bank_helpers_rounding.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    (1.999, 2, '2.00'),  # Test rounding up
+    (2.001, 2, '2.00'),  # Test rounding down
+    (1.555, 2, '1.56'),  # Test rounding halfway between two rounded values
+    (-1.5, 2, '-2.00'),  # Test rounding halfway between negative and positive values
+    (1234567890.123456789, 8, '1234567890.12345679'),  # Test rounding a float with a large number of digits before and after decimal
+    (987.654321, -2, '1000.00'),  # Test rounding a float with a negative number of digits after decimal
+    (42.123, 0, '42'),  # Test rounding a float with zero value for digits_after_decimal
+    (True, 2, '1.00'),  # Test rounding a non-numeric input
+    (10.00, 2, '10.00'),  # Test rounding a float that is already rounded to specified digits after decimal
+])
```
Comment (Bluesy1) [inline tests/test_problem_bank_helpers_rounding.py]: Consider setting the `ids` parameter for the parametrize mark, so if/when any of the tests here fail, you can see which semantic isn't operating as expected - see pytest docs [here](https://docs.pytest.org/en/7.3.x/example/parametrize.html#different-options-for-test-ids) for reference on setting parametrize ids

**Review by Bluesy1 — COMMENTED** [review]
I figured I should leave a few comments on this, so here goes:
This generally seems good to me, some minor suggestions that I think could improve the quality of the tests.
There's some tests that currently are testing multiple distinct things in a single test case - I think I commented on all of them, its a bad pattern for tests to test 2+ different uses of a single function in one test case (parametrized testing excluded where its multiple test cases using one function, but they get tested separately), we want to test all expected features every test run, even if a test fails.

One more suggestion - get this running in CI so other people can see tests run without having to clone the branch and run tests themself - you can copy the workflow file here and put it in `.github/workflows/tests.yml` - Just make sure you remove the `schema` job, since its not needed here.
https://github.com/open-resources/problem_bank_scripts/blob/main/.github/workflows/tests.yml

> The tests can be run from the root dir with:
> 
> ```
> pytest tests/
> ```

Also since this project uses poetry, you should be running the tests as `poetry run pytest tests`, to make sure its running against the editable install of the project made from running `poetry install`

**Review by Bluesy1 — APPROVED** [review]
It looks like you're going to need to run `poetry lock` and push the updated lock file - the dependencies specified in this packages lock are very out of date, ex numpy is on `1.24`, but the lock is for `1.20` still
Otherwise, it looks good to me!

@firasm thoughts?

**Comment by Bluesy1** [comment]
@amurph24 The dependencies are going to keep failing almost certainly until you tell poetry to update dependencie versions in the lock file - they're too old for modern github actions runners, unless you tell poetry to explicitly use the version of python you're setting up, which isn't worth keeping dependencies that will need to be updated soon anyways to support python versions used by prarielearn anyways

We did the same thing in problem_bank_scripts

**Comment by amurph24** [comment]
So far:
- tests have been created: maybe more tests are required to validate the csvs
- github CI has workflow testing functioning correctly
- minor changes to `problem_bank_helpers.py`, to help with default function calls and variables
## TODO:
- [x] ~~fix `sigfigs()` for scientific notation~~ left for a future pr, and ~~numbers with more than 9 sigfigs~~
- [x] fix `num_as_str()` for various cases that break, see failed tests
- [x] update `roundp()` and `round_str()` to correctly utilise `sigfig` package

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        run: |
+          poetry run pytest tests
+      - name: Upload generated files
+        if: failure()
+        uses: actions/upload-artifact@v3
+        with:
+          name: test-generated-files-${{ github.event_name }}-${{ matrix.python_version }}
+          path: tests/test_question_templates/question_generated_outputs/
+          if-no-files-found: ignore
+          retention-days: 5
```
Comment (Bluesy1) [inline .github/workflows/tests.yml]:
```suggestion

```
This one's on me for not mentioning - we don't need to upload generated files on test failures cause we never generate files in tests (for now at least)

File: numpy-1.20.3-cp38-cp38-win_amd64.whl
Comment (Bluesy1) [inline numpy-1.20.3-cp38-cp38-win_amd64.whl:1]: Is this wheel needed, or a leftover from debugging ci?

**Review by Bluesy1 — COMMENTED** [review]
I don't understand the code in the repo so I still can't comment on test quality, but a few minor things I noticed since the last time I looked at it that I felt I should mention
(Don't remove Firas' review request, these are small things that just were worth pointing out while I was going through all of my GH notifications from today 🙂)

File: numpy-1.20.3-cp38-cp38-win_amd64.whl
Comment (amurph24) [reply numpy-1.20.3-cp38-cp38-win_amd64.whl:1]: I think its a leftover from debugging, everything still works without it

**Comment by amurph24** [comment]
This should've been multiple PR's but I have most of the issues with the `pbh` functions resolved with the commits here, this also includes CI testing for `problem_bank_helpers`.

File: src/problem_bank_helpers/problem_bank_helpers.py [hunk 6]
```diff
@@ -77,7 +79,9 @@ def round_sig(x, sig):
         y = 0
     else:
         y = sig - int(floor(log10(abs(x)))) - 1
-    return round(x, y)
+    # avoid precision loss with floats 
+    x = Decimal(repr(x))
+    return float(round(x, y))
```
Comment (Bluesy1) [inline src/problem_bank_helpers/problem_bank_helpers.py:86]: Is there a reason decimal rounding context is set here (https://github.com/open-resources/problem_bank_helpers/pull/14/files#diff-245aa6c888f36f3441102ab89c24335637ea34f63e5d2f63beb583b866af41e1R245-R246), but not here?

File: tests/test_problem_bank_helpers_other_functionality.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+def test_empty_csv():
+    # ensure loading an empty csv fails
+    file_path = "data/empty.csv"
+
+    # Create an empty CSV file
+    with open(file_path, 'w') as csvfile:
+        pass
+
+    with pytest.raises(pd.errors.EmptyDataError):
+        test_load_csv(file_path)
```
Comment (Bluesy1) [inline tests/test_problem_bank_helpers_other_functionality.py:54]: I'd love some insight into these, after thinking about them again - are we just testing pandas here, or is this just to test that empty CSV files pass properly? Also if we do want to test the error on an empty csv, maybe we can use something like an in-memory tempfile using python's [tempfile stdlib module](https://docs.python.org/3.11/library/tempfile.html?highlight=tempfile#tempfile.SpooledTemporaryFile) instead of creating a file on disk.

File: tests/test_problem_bank_helpers_other_functionality.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import csv
+
+def test_version():
+    assert __version__ == '0.1.14'
+
+files = sorted(
+    [
+        file.name
+        for file in pathlib.Path(
+            "data/"
```
Comment (Bluesy1) [inline tests/test_problem_bank_helpers_other_functionality.py]:
```suggestion
            "src/problem_bank_helpers/data/"
```
This might require a rebase off of main, but the canonical location of the data files is no longer under the root directory

File: src/problem_bank_helpers/problem_bank_helpers.py [hunk 9]
```diff
@@ -225,13 +240,13 @@ def num_as_str(num, digits_after_decimal = 2):
...
     else:
-        from decimal import Decimal, getcontext, ROUND_HALF_UP
+        
 
         round_context = getcontext()
         round_context.rounding = ROUND_HALF_UP
 
-        tmp = Decimal(num).quantize(Decimal('1.'+'0'*digits_after_decimal))
-
+        tmp = Decimal(repr(num)).quantize(Decimal('1.'+'0'*digits_after_decimal))
```
Comment (Bluesy1) [inline src/problem_bank_helpers/problem_bank_helpers.py:250]: Any reason for using `repr` over `str` here?

File: tests/test_problem_bank_helpers_other_functionality.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        ).iterdir()
+    ]
+)
+
+files = [("data/" + f) for f in files if f != 'empty.csv']
```
Comment (Bluesy1) [inline tests/test_problem_bank_helpers_other_functionality.py]:
```suggestion
files = [("src/problem_bank_helpers/data/" + f) for f in files if f != 'empty.csv']
```
Same idea as my comment for line 18

File: tests/test_problem_bank_helpers_rounding.py [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+def test_num_as_str_default_dp():
+    """test default decimal places"""
+    assert pbh.num_as_str(3.14159) == '3.14'
+
+def test_num_as_str_invalid_args_kwargs():
+    with pytest.raises(TypeError):
+        pbh.num_as_str(123.45, 2, "args")  # Function should not accept *args
+
+    with pytest.raises(TypeError):
+        pbh.num_as_str(123.45, 2, digits_after_decimal=2)  # Function should not accept **kwargs
```
Comment (Bluesy1) [inline tests/test_problem_bank_helpers_rounding.py:202]: Perfectly valid test technically, just not sure the reason for this - this is more testing python itself than the function, since these errors will be raised from python itself from passing an invalid set of arguments to a function, over the function accepting a limited set of inputs to a `*args` and `**kwargs` allowed to be passed

(unless I'm missing something and `num_as_str` takes either/both of those, in which case please ignore)

File: pyproject.toml [hunk 12]
```diff
@@ -11,6 +11,7 @@ repository = "https://github.com/open-resources/problem_bank_helpers"
 include = [{ path = "data/", format = ["sdist", "wheel"]}]
 
 [tool.poetry.dependencies]
+
 python = "^3.9"
```
Comment (Bluesy1) [inline pyproject.toml]:
```suggestion
python = "^3.10"
```
Prairielearn runs on 3.10, so we should actually target that version of python to the best of our abilities (as much as its a moving target) - this one may require the lockfile being regenerated after to satisfy poetry, not sure

File: .github/workflows/tests.yml [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        uses: actions/setup-python@v3
+        with:
+          python-version: ${{ matrix.python_version }}
+          cache: poetry
+      - name: Install dependencies
+        run: |
+          poetry install
+      - name: Run tests
+        run: |
+          poetry run pytest tests
```
Comment (Bluesy1) [inline .github/workflows/tests.yml]:
```suggestion
    concurrency:
      group: tests-${{ github.ref }}
      cancel-in-progress: true
    steps:
      - uses: actions/checkout@v3
      - name: Install Poetry --with dev
        run: |
          curl -sSL https://install.python-poetry.org | python -
          echo "$HOME/.poetry/bin" >> $GITHUB_PATH
      - name: Set up Python 3.10
        uses: actions/setup-python@v4
        with:
          python-version: "3.10"
          cache: poetry
      - name: Install dependencies
        run: |
          poetry install
      - name: Run tests
        run: |
          poetry run pytest tests

```
Few minor tweaks to the actions config, as well as bumping python version as discussed in my comment on the `pyproject.toml`

File: src/problem_bank_helpers/problem_bank_helpers.py [hunk 14]
```diff
@@ -51,11 +52,12 @@ def sigfigs(x):
...
         myStr = x.split('e')
-        return len( myStr[0] ) - 1 # to compenstate for the decimal point
+        
+        return len( myStr[0] ) - (1 if '.' in x else 0) # to compenstate for the decimal point
     else:
         # put it in e format and return the result of that
-        ### NOTE: because of the 8 below, it may do crazy things when it parses 9 sigfigs
-        n = ('%.*e' %(8, float(x))).split('e')
+        ### NOTE: because of the 15 below, it may do crazy things when it parses 16 sigfigs
+        n = ('%.*e' %(15, float(x))).split('e')
```
Comment (Bluesy1) [inline src/problem_bank_helpers/problem_bank_helpers.py]:
```suggestion
        n = f'{float(x):.15e}'.split('e')
```
Seems more readable like this:
```python
>>> x = '1.234567890'
>>> ('%.*e' %(15, float(x))).split('e')
['1.234567890000000', '+00']
>>> f'{float(x):.15e}'.split('e')
['1.234567890000000', '+00']
```

**Review by Bluesy1 — COMMENTED** [review]
Seems pretty good overall, I'm not too worried about all the fixes going along with the tests, it makes some amount of sense to me, and its not in a singular commit here, so its not hard to pick through and see whats fixing/adding/changing what

I'd love to see all the new files have consistent newlines at EOF, the tests seem reasonable and functional with good coverage it looks like (unless we run the tests with `pytest-cov` installed, I can't say that for sure though), good job!

File: tests/test_problem_bank_helpers_rounding.py (hunk 11 above)
Comment (amurph24) [reply tests/test_problem_bank_helpers_rounding.py:202]: Nope very true, removed them.

File: .github/workflows/tests.yml [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+  merge_group:
+
+permissions:
+  contents: read
+
+jobs:
+  
+  tests:
+    runs-on: ubuntu-latest
+    strategy:
```
Comment (Bluesy1) [inline .github/workflows/tests.yml:32]:
```suggestion
```
The workflow file currently is invalid

**Comment by firasm** [comment]
@amurph24 - I'm merging this in, but I think some of the review comments from Gavin need to be addressed. Also, there are three warnings being produced:

```
▶ poetry run pytest
============================================================================================ test session starts =============================================================================================
platform darwin -- Python 3.10.2, pytest-6.2.5, py-1.11.0, pluggy-1.2.0
rootdir: /Users/firasm/Sync/EL/OPB/code/problem_bank_helpers
collected 111 items                                                                                                                                                                                          

tests/test_problem_bank_helpers.py .                                                                                                                                                                   [  0%]
tests/test_problem_bank_helpers_other_functionality.py ..............                                                                                                                                  [ 13%]
tests/test_problem_bank_helpers_rounding.py ................................................................................................                                                           [100%]

============================================================================================== warnings summary ==============================================================================================
tests/test_problem_bank_helpers_rounding.py::test_roundp_rigorous_sigfig[id_zero float---]
tests/test_problem_bank_helpers_rounding.py::test_roundstr_rigorous_sigfig[id_zero float---]
  /Users/firasm/Library/Caches/pypoetry/virtualenvs/problem-bank-helpers-NBE3jFb2-py3.10/lib/python3.10/site-packages/sigfig/sigfig.py:586: UserWarning: warning: 3 significant figures requested from number with only 1 significant figures
    warn("warning: %d significant figures requested from number with only %d significant figures" % (given['sigfigs'], len(num.map)))

tests/test_problem_bank_helpers_rounding.py::test_roundstr_with_decimal_sci_notation_dp
  /Users/firasm/Library/Caches/pypoetry/virtualenvs/problem-bank-helpers-NBE3jFb2-py3.10/lib/python3.10/site-packages/sigfig/sigfig.py:391: UserWarning: Cannot round by both sigfigs & decimals, ignoring decimal constraint
    warn('Cannot round by both sigfigs & decimals, ignoring decimal constraint')

-- Docs: https://docs.pytest.org/en/stable/warnings.html
====================================================================================== 111 passed, 3 warnings in 2.17s =======================================================================================
```

File: src/problem_bank_helpers/problem_bank_helpers.py (hunk 9 above)
Comment (amurph24) [reply src/problem_bank_helpers/problem_bank_helpers.py:250]: nope, it has been addressed now

File: tests/test_problem_bank_helpers_other_functionality.py (hunk 7 above)
Comment (amurph24) [reply tests/test_problem_bank_helpers_other_functionality.py:54]: These tests are not super clear, and do need to be reworked or improved substaintially.
- `test_load_csv` tests that the files can be loaded as pandas dataframes.
- `test_empty_csv` tests that empty CSV's will fail; the intention was to make sure empty CSV files fail appropriately, and validate that `test_load_csv` is working correctly and will not pass an empty CSV. As you mentioned this essentially is testing that pandas will throw an error if you try to load an empty csv, so perhaps this should be removed.

We definitely should be using a tempfile instead of storing an empty csv.

File: src/problem_bank_helpers/problem_bank_helpers.py (hunk 6 above)
Comment (amurph24) [reply src/problem_bank_helpers/problem_bank_helpers.py:86]: no reason, context is now set at the top of the file
